### PR TITLE
Issue1940 dev docu

### LINKED ISSUE #1940: Update development documentation
Based on feedback from new developers and a PR checklist draft google doc, and some currently standing needs:

* convert to `rst` and move the content of https://github.com/lbl-srg/modelica-buildings/wiki/Style-Guide to the user guide, section after 5.2 in https://simulationresearch.lbl.gov/modelica/userGuide/development.html
* integrate the [PR checklist google doc items](https://docs.google.com/document/d/13eNyQkj628dWBaruJ2wT8xYCfUkSCZiTmksSfYB-Uck/edit?usp=sharing)
* account for naming conventions that were recently added

### PR DESCRIPTION
**Pull request by milicag**
Closes #1940. Integrates the style guide wiki page and aims to replace it. Integrates the PR checklist created by an array of control sequence developers.

### CODE REVIEW

**Comment by milicag** [comment]
@mwetter The test fails on the Sphinx version. Please comment on the the edits to the content when you get a chance and I will use the aligned version to recreate the `html`s.

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 1]
```diff
@@ -37,6 +37,377 @@ Models that are contributed need to adhere to the following guidelines, as this
...
+#. Make sure that the derivatives of equations are bounded on compact
+   sets. For example, instead of using ``y=sign(x) * sqrt(abs(x))``,
+   approximate the equation with a differentiable function that has a
+   finite derivative near zero. Use functions form
+   ``Buildings.Utilities.Math`` for this approximation.
+#. Whenever possible, a Modelica tool should not have to do numerical
+   differentiation. For example, in Dymola, if your model translation
+   log shows
+
+   ::
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:185]: This should be rendered as ..code::

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 2]
```diff
@@ -37,6 +37,377 @@ Models that are contributed need to adhere to the following guidelines, as this
...
+
+       Number of numerical Jacobians: 1
+
+   (or any number other than zero), then enter on the command line
+
+   ::
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:191]: this should be rendered as ..code::

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 3]
```diff
@@ -37,6 +37,377 @@ Models that are contributed need to adhere to the following guidelines, as this
...
+
+   Next, translate the model again to see what functions cannot be
+   differentiated symbolically. Then, implement symbolic derivatives for
+   this function.
+   See `implementation of function derivatives <Function-Derivatives>`__.
+
+
+Functions
+~~~~~~~~~
+
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:204]: One blank line suffices.

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 4]
```diff
@@ -170,18 +539,38 @@ The following rules need to be followed, in addition to the guidelines described
    Do not use ``Dialog(tab=STRING))``, unless the parameter is declared with a default value
    and this parameter and its value is of no interest to typical users.
 
-#. Each block must have a ``defaultComponentName`` annotation.
+#. In the code text the instances must be ordered as follows:
+     - If an instance has a type associated with it, then list them in this order: `boolean`, `integer`, `real`.
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:543]: I don't understand this. Every instance has a type. Also, use the Modelica types, e.g., `Real` and not `real`.

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 5]
```diff
@@ -170,18 +539,38 @@ The following rules need to be followed, in addition to the guidelines described
...
+     - Instances of lower dimensionality should come first, e.g. scalar values before arrays, 
+       though grouping based on model specific similarities may be prioritized
+     - Instances should be ordered as: parameters, interfaces: first inputs, then outputs, then ports,
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:546]: There are no `ports` in CDL  and simply call them inputs and outputs, not interfaces as interface is not defined what it is. Also, correct punctuation to use a full sentence.

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 6]
```diff
@@ -170,18 +539,38 @@ The following rules need to be followed, in addition to the guidelines described
...
+       followed by any blocks/instances
+     - Protected instances are below all the unprotected instances and follow the same instance ordering rules.
+
+#. Each block must have a ``defaultComponentName`` annotation and a ``%name`` at icon layer
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:550]: Replace with "and a ``%name`` label placed above the icon.

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 7]
```diff
@@ -170,18 +539,38 @@ The following rules need to be followed, in addition to the guidelines described
...
-   assign each parameter value on a new line, list first all parameters, then all inputs, all
-   outputs, and finally all instances of other blocks.
+   assign each parameter value on a new line. It is recommended to add an empty line between instances.
    See for example
    `G36_PR1.AHUs.SingleZone.VAV.SetPoints.ExhaustDamper <https://github.com/lbl-srg/modelica-buildings/blob/94d5919dbe1b2f2e317e7b69800f3b3ad07be930/Buildings/Controls/OBC/ASHRAE/G36_PR1/AHUs/SingleZone/VAV/SetPoints/ExhaustDamper.mo>`_.
 
 #. For parameters, where generally valid values can be provided, provide them
    as default values.
 
+#. Note that the protected instances are not accessible with the plots used in the validation tests.
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:561]: This does not belong here; let's not mix how tools work with how to implement sequences.

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 8]
```diff
@@ -170,18 +539,38 @@ The following rules need to be followed, in addition to the guidelines described
...
+
+#. Add comments to all the instances such as parameters, interfaces, blocks. The comments should be short and concise. The comments
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:563]: There are no interfaces, and "concise" is by definition short.

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 9]
```diff
@@ -170,18 +539,38 @@ The following rules need to be followed, in addition to the guidelines described
...
 
 #. For parameters, where generally valid values can be provided, provide them
    as default values.
 
+#. Note that the protected instances are not accessible with the plots used in the validation tests.
+
+#. Add comments to all the instances such as parameters, interfaces, blocks. The comments should be short and concise. The comments 
+   should not contain redundant information and hard coded parameters. If the functionality of an instance is obvious the developer may use
+   comments that closely resemble the class names, such as `Logical And`. 
+   Make sure comments are short and don’t contain any redundant information.
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:566]: This is said on line 563 already.

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 10]
```diff
@@ -219,25 +608,29 @@ The following rules need to be followed, in addition to the guidelines described
 
     Buildings.Controls.OBC.CDL.Interfaces.RealInput TZon(
       final unit="K",
-      displayUnit="degC") "Measured zone air temperature";
+      final displayUnit="degC") "Measured zone air temperature";
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:611]: Display unit must not be final. Users may want to change this to IP.

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 11]
```diff
@@ -219,25 +608,29 @@ The following rules need to be followed, in addition to the guidelines described
...
       final unit="1") "Exhaust damper position";
 
-   Here, units are declared as ``final`` to avoid users to be able to change them, as
-   a change in unit may cause the control logic to be incorrect.
-
    Conversion of these units to non-SI units can be done programmatically by tools that
    process CDL.
 
+#. Units, quantities and value limits must be declared as ``final`` to avoid users to be able to change them, as
+   a change in unit may cause the control logic to be incorrect. Often display units are also declared as final.
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:625]: remove " Often display units are also declared as final."

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 12]
```diff
@@ -249,22 +642,46 @@ The following rules need to be followed, in addition to the guidelines described
    Make sure sequences are tested for all modes of operation, and as applicable, for winter, shoulder
    and summer days.
 
+#. Perform code text and, if applicable, figure inspection to:
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:645]: Remove to be concise. It should be clear that code and figures need to be inspected.

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 13]
```diff
@@ -249,22 +642,46 @@ The following rules need to be followed, in addition to the guidelines described
...
+      - ensure there are no remaining typos
+      - ensure no element `public` was added automatically after the `protected` code section
 
+#. Use full paths to blocks, that is paths that start with `Buildings.Controls...`
 
-.. _sec_val:
+#. Final steps for a pull requests:
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:651]: Remove. This belongs to how to make a pull request. It is not specific to control sequences.

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 14]
```diff
@@ -249,22 +642,46 @@ The following rules need to be followed, in addition to the guidelines described
...
+      - make sure that the latest base branch is merged into the development branch
+      - after creating a pull request observe that the files changed pertain to the 
+        issue being addressed with the development branch
 
-Validation
-----------
+#. For general rules on validation models see :numref:`sec_val`. Specifically for the 
+   sequence development ensure that:
+      - the instances of models being tested are listed at the top of the code text
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:658]: I don't understand this line.

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 15]
```diff
@@ -249,22 +642,46 @@ The following rules need to be followed, in addition to the guidelines described
...
+      - add a blank line at the end of each .mos script
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:659]: Why is a blank line needed? If it is needed, is it a bug in BuildingsPy? If it is needed for Dymola, it belongs to the description of how to implement unit tests. This needs to be about control sequences only. If we start mixing it, we will repeat ourself and won't be concise.

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 16]
```diff
@@ -249,22 +642,46 @@ The following rules need to be followed, in addition to the guidelines described
...
 
-All models that are implemented need to be validated for all
-realistic operating modes.
-These validations need to be part of the
-`unit tests <https://github.com/lbl-srg/modelica-buildings/wiki/Unit-Tests>`_.
+#. Run modelica-json$ node app.js -f Buildings/Controls/OBC/ASHRAE/PrimarySystem/{path to package} -o json -m modelica
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:661]: Format as code.

File: Buildings/Resources/Documentation/userGuide/source/development.rst [hunk 17]
```diff
@@ -249,22 +642,46 @@ The following rules need to be followed, in addition to the guidelines described
...
+-------------------------
+
+The developer that introduces a new model, block or a function must:
+
+1. Implement at least one example or validation model that serves as a unit test for each model, block and function,
+   and run the unit tests.
+   Unit tests should cover all branches of ``if-then`` constructs and 
+   all realistic operating modes of the system represented by the model.
+   See `unit test implementation <Unit-Tests>`__.
+2. Ensure that no unit test requires a numerical Jacobian. If a
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/source/development.rst:676]: This is already explained on line 187, remove it here.

File: Buildings/Resources/Documentation/userGuide/build/html/workArounds.html [hunk 18]
```diff
@@ -322,7 +329,7 @@ <h2>3.5. Avoiding overspecified initialization problems<a class="headerlink" hre
     </p>
     <p>
         &copy; Copyright 2012-2020, The Regents of the University of California (through Lawrence Berkeley National Laboratory).<br/>
-      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.1.2.<br/>
+      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.<br/>
```
Comment (mwetter) [inline Buildings/Resources/Documentation/userGuide/build/html/workArounds.html:332]: It looks like you have an old version of sphinx.

File: Buildings/Resources/Documentation/userGuide/source/development.rst (hunk 5 above)
Comment (milicag) [reply Buildings/Resources/Documentation/userGuide/source/development.rst:546]: `ports` is legacy from previous version of the document, as it is for both CDL and other development within the library.

File: Buildings/Resources/Documentation/userGuide/source/development.rst (hunk 4 above)
Comment (milicag) [reply Buildings/Resources/Documentation/userGuide/source/development.rst:543]: An instance of a subsequence block does not have a type.

File: Buildings/Resources/Documentation/userGuide/source/development.rst (hunk 14 above)
Comment (milicag) [reply Buildings/Resources/Documentation/userGuide/source/development.rst:658]: If I am testing a MZ VAV controller then an instance of it should come before any other block instances included in the test.

File: Buildings/Resources/Documentation/userGuide/source/development.rst (hunk 15 above)
Comment (milicag) [reply Buildings/Resources/Documentation/userGuide/source/development.rst:659]: I assume it was added due to some tests failing if there isn't a blank line at the end of a mos file.

File: Buildings/Resources/Documentation/userGuide/build/html/workArounds.html (hunk 18 above)
Comment (milicag) [reply Buildings/Resources/Documentation/userGuide/build/html/workArounds.html:332]: planning to update

File: Buildings/Resources/Documentation/userGuide/source/development.rst (hunk 7 above)
Comment (milicag) [reply Buildings/Resources/Documentation/userGuide/source/development.rst:561]: Is there a better spot for it? Maybe in the testing section?

File: Buildings/Resources/Documentation/userGuide/source/development.rst (hunk 10 above)
Comment (milicag) [reply Buildings/Resources/Documentation/userGuide/source/development.rst:611]: That makes perfect sense - however we've been using final with the displayUnit throughout the OBC package. It should be easy to replace.

File: Buildings/Resources/Documentation/userGuide/source/development.rst (hunk 7 above)
Comment (milicag) [reply Buildings/Resources/Documentation/userGuide/source/development.rst:561]: I moved it to the end of the file.

File: Buildings/Resources/Documentation/userGuide/source/development.rst (hunk 8 above)
Comment (milicag) [reply Buildings/Resources/Documentation/userGuide/source/development.rst:563]: Agreed

File: Buildings/Resources/Documentation/userGuide/source/development.rst (hunk 14 above)
Comment (milicag) [reply Buildings/Resources/Documentation/userGuide/source/development.rst:658]: It is our own recommendation, similarly to ordering of types or instances.

File: Buildings/Resources/Documentation/userGuide/source/development.rst (hunk 15 above)
Comment (milicag) [reply Buildings/Resources/Documentation/userGuide/source/development.rst:659]: I'll move these two statements to the unit testing wiki page:

* Protected instances are not accessible through `mos` scripts. 
* A blank line at the end of each .mos script

What's confusing is that we have a Validation and Unit Tests section in this file too.

Comment (milicag) [reply Buildings/Resources/Documentation/userGuide/source/development.rst]: Moved.

File: Buildings/Resources/Documentation/userGuide/source/development.rst (hunk 14 above)
Comment (milicag) [reply Buildings/Resources/Documentation/userGuide/source/development.rst]: I rephrased, see if it works

File: Buildings/Resources/Documentation/userGuide/source/development.rst (hunk 7 above)
Comment (milicag) [reply Buildings/Resources/Documentation/userGuide/source/development.rst]: Moved to unit test wiki page.

File: Buildings/Resources/Documentation/userGuide/source/development.rst (hunk 10 above)
Comment (milicag) [reply Buildings/Resources/Documentation/userGuide/source/development.rst:589]: #2045

**Comment by milicag** [comment]
@mwetter I've addressed your comments and updated my Sphinx. 5.4.3 under 3. and 17. are formatted bold in the html and I don't see directly why. Please take a look again and let me know of any remaining issues.

**Comment by mwetter** [comment]
@milicag : Please wait with further edits as I am reviewing it now.

**Comment by mwetter** [comment]
@milicag : Do we really need these two statements below? I think the first is not needed because a Modelica tool takes care of this, and `modelica-json` should as well, and I am not sure whether the second is really sufficiently important.
```
#. Use full paths to blocks, that is paths that start with ``Buildings.Controls...``.
```

```
#. For general rules on validation models see :numref:`sec_val`. Preferably,
   list the block being tested at the top of the code, and list inputs to the block afterwards.
```
Once these are addressed and all tests pass, it is ready to merge. Then we should also delete https://github.com/lbl-srg/modelica-buildings/wiki/Style-Guide. I made the changes also to https://github.com/ibpsa/modelica-ibpsa/wiki/Style-Guide so that they are consistent.

**Comment by milicag** [comment]
@mwetter 

first: If there is really no concern about the first one I'm happy to delete it.

second: it really isn't a must, I got used to that structure similarly to say order of types: Boolean, Integer, Real. At the very least I feel that the tested instances should be grouped together, such that we don't have one tested instance, then a bunch of say pulse signals, than a constant, then again tested instances. Up to you.

File: Buildings/Resources/Documentation/userGuide/build/html/_sources/development.rst.txt [hunk 19]
```diff
@@ -619,7 +619,7 @@ The following rules need to be followed, in addition to the guidelines described
 
 #. The size of the icon should be such that it provides a good fit for all the input and output interfaces. The minimum
    recommended icon size is 100 by a 100. If there are many interfaces the icon size should be extended in vertical direction.
-   Icons should be symetrical with reference to the grid origin. E.g, the default specification is
+   Icons should be symmetrical with reference to the grid origin. E.g, the default specification is
```
Comment (milicag) [inline Buildings/Resources/Documentation/userGuide/build/html/_sources/development.rst.txt:622]: ups, I forgot the spell check isn't active as I've changed text editors. Apologies!

**Pull request merged**